rust/hyperlane-base/src/types/abs_storage.rs

```diff
@@ -0,0 +1,202 @@
+//! Azure Blob Storage provider for `CheckpointSyncer`.
+use super::{ANNOUNCEMENT_KEY, LATEST_INDEX_KEY};
+use crate::CheckpointSyncer;
+use async_trait::async_trait;
+use azure_storage::prelude::*;
+use azure_storage_blobs::prelude::*;
+use derive_new::new;
+use eyre::{bail, Result};
+use hyperlane_core::{SignedAnnouncement, SignedCheckpointWithMessageId};
+use std::fmt;
+use time::Duration;
+
+/// Azure blob storage user env variable
+pub const ABS_USER: &str = "ABS_USER";
+/// Azure blob storage access key env variable
+pub const ABS_ACCESS_KEY: &str = "ABS_ACCESS_KEY";
+
+/// Builder for [AbsStorageClient]
+/// Provide `(None, None)` for no-auth anonymous access to public container
+/// Use `.build()` method with container and blob to finalize instantiation
+#[derive(Debug, new)]
+pub struct AbsStorageClientBuilder {
+    /// Account name.
+    ///
+    /// If this and [access_key] are not provided, anonymous access is assumed for public container read.
+    account: Option<String>,
+
+    /// Access Key for given account.
+    ///
+    /// If this and [account] are not provided, anonymous access is assumed for public container read.
+    access_key: Option<String>,
+}
+
+/// Azure Blob Storage client
+/// Instantiated for a single container and single blob (file)
+/// Supports both Authentitated (read/write) and Anonymous (read) modes
+/// Use [AbsStorageClientBuilder] to configure
+pub struct AbsStorageClient {
+    // ABS storage client
+    inner: ClientBuilder,
+    // container
+    container: String,
+}
+
+impl AbsStorageClientBuilder {
+    /// Instantiates `AbsStorageClient` based on provided auth method
+    /// # Param
+    /// * `container` - String'ish name of target container to work with, will be used by all store and get ops
+    /// blob names are either defined by constants from 'super' or are build using `get_checkpoint_key(id)` function
+    pub fn build(self, container: impl Into<String>) -> AbsStorageClient {
+        let (account, credential) = if let Some((account, key)) = self.into_account_and_key() {
+            (
+                account.clone(),
+                StorageCredentials::access_key(account, key),
+            )
+        } else {
+            ("".into(), StorageCredentials::anonymous())
+        };
+        AbsStorageClient {
+            inner: ClientBuilder::new(account, credential),
+            container: container.into(),
+        }
+    }
+
+    fn into_account_and_key(self) -> Option<(String, String)> {
+        if self.account.is_some() && self.access_key.is_some() {
+            Some((self.account.unwrap(), self.access_key.unwrap()))
+        } else {
+            None
+        }
```

Comment:
nit - more idiomatic to do as below. Though if the usage pattern requires both `account` and `access_key` at all times, I'd store them both as a tuple in `self`, maybe under a field called `credentials` that points to a struct with these two fields
```suggestion
        match (self.account, self.access_key) {
            (Some(account), Some(access_key)) => Some((account, access_key)),
            _ => None,
        }
```